File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example/h2x/low_level/lrcu_cmd.cpp
/*
 * 注意事项
 *
 * 1. 机器人模式切换需要确保当前状态安全
 * 2. 移动命令需要在合适的模式下执行
 * 3. 所有操作结果通过result字段返回，0表示成功，非0表示错误
 */

// lrcu_cmd.cpp

#include <iostream>
#include <thread>
#include <chrono>

#include "common/common_def.h"
#include "common/h2x_def.h"
#include "h2x/h2x_low_level.h"

using namespace humanoid;
using namespace humanoid::h2x;

int main() {
  std::cout << "=== LRCU Command Test ===" << std::endl;
  
  H2xLowLevel low_level;
  LRCUCmd lrcu_cmd;

  // Test LED control for 10 seconds
  constexpr int DURATION_MS = 10000;
  constexpr int CMD_INTERVAL_MS = 500;
  constexpr int cmd_count = DURATION_MS / CMD_INTERVAL_MS;

  for (int i = 0; i < cmd_count; ++i) {
    auto start = std::chrono::high_resolution_clock::now();

    // Set LED parameters
    lrcu_cmd.led_enable = 1;  // Enable LED
    lrcu_cmd.led_period_ms = 1000;  // 1 second period

    // Cycle through different colors
    int color_phase = i % 6;
    switch (color_phase) {
      case 0:  // Red
        lrcu_cmd.led_r = 255;
        lrcu_cmd.led_g = 0;
        lrcu_cmd.led_b = 0;
        lrcu_cmd.led_mode = 0;  // Solid color
        std::cout << "Setting LED to RED" << std::endl;
        break;
      case 1:  // Green
        lrcu_cmd.led_r = 0;
        lrcu_cmd.led_g = 255;
        lrcu_cmd.led_b = 0;
        lrcu_cmd.led_mode = 0;
        std::cout << "Setting LED to GREEN" << std::endl;
        break;
      case 2:  // Blue
        lrcu_cmd.led_r = 0;
        lrcu_cmd.led_g = 0;
        lrcu_cmd.led_b = 255;
        lrcu_cmd.led_mode = 0;
        std::cout << "Setting LED to BLUE" << std::endl;
        break;
      case 3:  // Yellow
        lrcu_cmd.led_r = 255;
        lrcu_cmd.led_g = 255;
        lrcu_cmd.led_b = 0;
        lrcu_cmd.led_mode = 1;  // Blinking
        std::cout << "Setting LED to YELLOW (Blinking)" << std::endl;
        break;
      case 4:  // Cyan
        lrcu_cmd.led_r = 0;
        lrcu_cmd.led_g = 255;
        lrcu_cmd.led_b = 255;
        lrcu_cmd.led_mode = 1;
        std::cout << "Setting LED to CYAN (Blinking)" << std::endl;
        break;
      case 5:  // Magenta
        lrcu_cmd.led_r = 255;
        lrcu_cmd.led_g = 0;
        lrcu_cmd.led_b = 255;
        lrcu_cmd.led_mode = 1;
        std::cout << "Setting LED to MAGENTA (Blinking)" << std::endl;
        break;
    }

    // Send LRCU command
    int ret = low_level.writeLRCUCmd(lrcu_cmd);
    if (ret != 0) {
      std::cout << "Failed to write LRCU command!" << std::endl;
      return -1;
    }

    auto end = std::chrono::high_resolution_clock::now();
    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    std::this_thread::sleep_for(std::chrono::milliseconds(CMD_INTERVAL_MS) - elapsed);
  }

  // Turn off LED at the end
  lrcu_cmd.led_enable = 0;
  low_level.writeLRCUCmd(lrcu_cmd);
  std::cout << "LED turned off" << std::endl;

  std::cout << "LRCU command test completed successfully" << std::endl;
  return 0;
}

File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example_py/h2x/low_level/lrcu_cmd.py
#!/usr/bin/env python3
#!
#!注意事项
#!
#!1. 机器人模式切换需要确保当前状态安全
#!2. 移动命令需要在合适的模式下执行
#!3. 所有操作结果通过result字段返回，0表示成功，非0表示错误


import humanoid_sdk_py
from humanoid_sdk_py import h2x
import time


def main():
    print("=== LRCU Command Test ===")

    # Initialize low level object
    low_level = h2x.H2xLowLevel()

    # Test LED control for 10 seconds
    DURATION_MS = 10000
    CMD_INTERVAL_MS = 500
    cmd_count = DURATION_MS // CMD_INTERVAL_MS

    for i in range(cmd_count):
        start_time = time.time()

        # Create LRCU command
        lrcu_cmd = h2x.LRCUCmd()
        
        # Set LED parameters
        lrcu_cmd.led_enable = 1  # Enable LED
        lrcu_cmd.led_period_ms = 1000  # 1 second period

        # Cycle through different colors
        color_phase = i % 6
        if color_phase == 0:  # Red
            lrcu_cmd.led_r = 255
            lrcu_cmd.led_g = 0
            lrcu_cmd.led_b = 0
            lrcu_cmd.led_mode = 0  # Solid color
            print("Setting LED to RED")
        elif color_phase == 1:  # Green
            lrcu_cmd.led_r = 0
            lrcu_cmd.led_g = 255
            lrcu_cmd.led_b = 0
            lrcu_cmd.led_mode = 0
            print("Setting LED to GREEN")
        elif color_phase == 2:  # Blue
            lrcu_cmd.led_r = 0
            lrcu_cmd.led_g = 0
            lrcu_cmd.led_b = 255
            lrcu_cmd.led_mode = 0
            print("Setting LED to BLUE")
        elif color_phase == 3:  # Yellow
            lrcu_cmd.led_r = 255
            lrcu_cmd.led_g = 255
            lrcu_cmd.led_b = 0
            lrcu_cmd.led_mode = 1  # Blinking
            print("Setting LED to YELLOW (Blinking)")
        elif color_phase == 4:  # Cyan
            lrcu_cmd.led_r = 0
            lrcu_cmd.led_g = 255
            lrcu_cmd.led_b = 255
            lrcu_cmd.led_mode = 1
            print("Setting LED to CYAN (Blinking)")
        elif color_phase == 5:  # Magenta
            lrcu_cmd.led_r = 255
            lrcu_cmd.led_g = 0
            lrcu_cmd.led_b = 255
            lrcu_cmd.led_mode = 1
            print("Setting LED to MAGENTA (Blinking)")

        # Send LRCU command
        ret = low_level.writeLRCUCmd(lrcu_cmd)
        if ret != 0:
            print("Failed to write LRCU command!")
            return -1

        elapsed_time = (time.time() - start_time) * 1000
        sleep_time = CMD_INTERVAL_MS - elapsed_time
        if sleep_time > 0:
            time.sleep(sleep_time / 1000.0)

    # Turn off LED at the end
    lrcu_cmd.led_enable = 0
    low_level.writeLRCUCmd(lrcu_cmd)
    print("LED turned off")

    print("LRCU command test completed successfully")
    return 0


if __name__ == "__main__":
    exit(main())